share/dkg/dkg.go

```diff
@@ -0,0 +1,611 @@
+		return err
+	}
+
+	var arr = d.pendingReconstruct[rs.DealerIndex]
+	// check if packet is already received or not
+	for _, r := range arr {
+		if r.Index == rs.Index {
+			// no-op ? error ?
+			return nil
+		}
+	}
+	// add it to list of pending
+	arr = append(arr, rs)
+	d.pendingReconstruct[rs.DealerIndex] = arr
+	// check if we can reconstruct commitments
+	if len(arr) >= d.t {
+		var shares = make([]*share.PriShare, 0, len(arr))
+		for _, r := range arr {
+			shares = append(shares, r.Share)
+		}
+		// error only happens when you have less than t shares, but we ensure
+		// there are more just before
+		pri, _ := share.RecoverPriPoly(d.suite, shares, d.t, len(d.participants))
+		d.commitments[rs.DealerIndex] = pri.Commit(d.suite.Point().Base())
+		// note it has been reconstructed.
+		d.reconstructed[rs.DealerIndex] = true
+		delete(d.pendingReconstruct, rs.DealerIndex)
+	}
+	return nil
+}
+
+// Finished returns true if the DKG has operated the protocol correctly and has
+// all necessary information to generate the DistKeyShare() by itself. It
+// returns false otherwise.
+func (d *DistKeyGenerator) Finished() bool {
+	var ret = true
+	var nb = 0
+	d.qualIter(func(idx uint32, v *vss.Verifier) bool {
+		nb++
+		// ALL QUAL members should have their commitments by now either given or
+		// reconstructed.
+		if _, ok := d.commitments[idx]; !ok {
+			ret = false
+			return false
+		}
+		return true
+	})
+	return nb >= d.t && ret
+}
+
+// ProduceSharedSecret generates the sharedsecret relative to this receiver
+// It throws an error if something is wrong such as not enough deals received.
+// The shared secret can be computed when all deals have been sent and
+// basically consists of a
+// 1. Public point which is the sum of all aggregated individual public commits
+// of each individual secrets.
+// 2. Share of the global Private Polynomial, basically SUM of fj(i) for a receiver i
+func (d *DistKeyGenerator) DistKeyShare() (*DistKeyShare, error) {
+	if !d.Certified() {
+		return nil, errors.New("dkg: distributed key not certified")
+	}
+
+	sh := d.suite.Scalar().Zero()
+	var pub *share.PubPoly
+	var err error
+
+	d.qualIter(func(i uint32, v *vss.Verifier) bool {
+		// share of dist. secret = sum of all share received.
+		s := v.Deal().SecShare.V
+		sh = sh.Add(sh, s)
+		// Dist. public key = sum of all revealed commitments
+		poly, ok := d.commitments[i]
+		if !ok {
+			err = fmt.Errorf("dkg: protocol not finished: %d commitments missing", i)
+			return false
+		}
+		if pub == nil {
+			// first polynomial we see (instead of generating n empty commits)
+			pub = poly
+			return true
+		}
+		pub, err = pub.Add(poly)
+		if err != nil {
+			return false
+		}
+		return true
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	_, commits := pub.Info()
+
+	return &DistKeyShare{
+		Commits: commits,
+		Share: &share.PriShare{
+			I: int(d.index),
+			V: sh,
+		},
+	}, nil
+}
+
+func (sc *SecretCommits) Hash(s abstract.Suite) []byte {
```

Comment:
godoc: Why does this pass `golint`? There should be a comment, no? Same for next two `Hash`s.